3.13/redundancy_eliminator.md

```diff
@@ -71,17 +71,29 @@ The hiearchy of types in the type system can also be represented
 by a lattice. Roughly, it looks like this:
```

Comment:
What understanding to we gain by saying this is a lattice, rather than a DAG? It doesn't seem to be at all related to the lattice of the variables and stack items (where I actually do understand that "lattice" is the right concept).

Reply:
Probably have to ask @markshannon for this one.

Reply:
Oh I remembered why, the edges aren't directional but bi-directional, and there should be edges going from each type to every other type, but that is omitted for the sake of clarity. So it's indeed a lattice.

Reply:
Okay, then I am even more confused. What do edges represent? If we have a node that is True, what does an edge to False mean? That the initial knowledge was a lie?

Reply:
Woops I misspoke. This is the correct version:

> there should be edges going from each type to ~every other type~ **the top**

Also I went to refresh my knowledge on lattices, this is a good read https://www.infinitelymore.xyz/p/lattices. The edges represent a partial order relation (like in a Hasse diagram).

Reply:
The latest commit tries to explain this further. Hopefully it aids your understanding!

Reply:
Thanks! I think my problem was that I know this structure very well, but I've never thought of it as a lattice -- usually I think of it as a DAG where edges represent a subtype relationship, and sometimes I think of it as a collection of sets where each node represents a strict superset of all of the nodes it links to -- both are just different ways to say "partial ordering" (of sets of values). :-)

It just occurred to me that in this diagram, when a node has multiple outgoing edges, there are two possibly interesting properties of our DAG (but not necessarily of all DAGs):

1. The nodes pointed to by edges going out of a given node all represent disjoint sets
2. The union of all nodes pointed to by a given node's outgoing edges *may* be equal to the source node

I don't think either of those properties is all that useful though. (1) doesn't have to hold, and (2) doesn't hold when the outgoing edge goes to _top_, nor presumably in _other constants_ (but it holds for all other nodes in this particular diagram). It's possible that (1) is useful when we're replacing weaker guards with stronger ones.